File: lectures/lecture-03-why-the-repository-must-become-the-system-of-record/exercises/exercise-01-fresh-session-answers/starter/typescript/main.ts
// fresh-session-answers exercise, TypeScript starter.
//
// All five questions are attempted, but three extractors are naive first
// drafts with a realistic mistake each (see SPEC.md "Starter state"):
// how-organized answers from the instructions file instead of the
// architecture doc, how-to-verify grabs the first line that MENTIONS
// verification instead of the Verification line, and where-are-we returns
// the progress file's heading instead of the Next best step line. Fix the
// three per SPEC.md. Run ../../verify.sh --stack=typescript until it
// exits 0.

import { existsSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import { pathToFileURL } from "node:url";

interface Question {
  readonly id: string;
  readonly question: string;
  readonly answered: boolean;
  readonly answer: string | null;
  readonly source: string | null;
}

function firstProseLine(path: string): string | null {
  if (!existsSync(path)) return null;
  for (const line of readFileSync(path, "utf8").split(/\r?\n/)) {
    const stripped = line.trim();
    if (stripped && !stripped.startsWith("#")) return stripped;
  }
  return null;
}

function lineMentioning(path: string, needle: string): string | null {
  // Naive helper: first line whose text contains the needle (any case).
  if (!existsSync(path)) return null;
  for (const line of readFileSync(path, "utf8").split(/\r?\n/)) {
    if (line.toLowerCase().includes(needle) && line.trim()) return line.trim();
  }
  return null;
}

function rawFirstLine(path: string): string | null {
  // Naive helper: first non-empty line, headings included.
  if (!existsSync(path)) return null;
  for (const line of readFileSync(path, "utf8").split(/\r?\n/)) {
    if (line.trim()) return line.trim();
  }
  return null;
}

function taggedLine(path: string, tag: string): string | null {
  if (!existsSync(path)) return null;
  const escaped = tag.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const match = readFileSync(path, "utf8").match(new RegExp(`^- ${escaped}: (.+)$`, "m"));
  return match && match[1] ? match[1].trim() : null;
}

function instructionsFile(repo: string): string | null {
  for (const name of ["AGENTS.md", "CLAUDE.md"]) {
    if (existsSync(join(repo, name))) return name;
  }
  return null;
}

function question(
  id: string,
  text: string,
  answer: string | null,
  source: string | null,
): Question {
  return {
    id,
    question: text,
    answered: answer !== null,
    answer,
    source: answer !== null ? source : null,
  };
}

export function readRepo(repo: string): Record<string, unknown> {
  const entryName = instructionsFile(repo);
  const entryPath = entryName ? join(repo, entryName) : null;

  const questions: Question[] = [
    question(
      "what-is-this", "What is this system?",
      entryPath ? firstProseLine(entryPath) : null,
      entryName,
    ),
    // Naive draft: answers "how is it organized" from the instructions
    // file's overview line. Exercise: extract the first prose line of
    // docs/ARCHITECTURE.md, source "docs/ARCHITECTURE.md".
    question(
      "how-organized", "How is it organized?",
      entryPath ? firstProseLine(entryPath) : null,
      entryName,
    ),
    question(
      "how-to-run", "How do I run it?",
      entryPath ? taggedLine(entryPath, "Run") : null,
      entryName ? `${entryName} (Run line)` : null,
    ),
    // Naive draft: any line that mentions verification. Prose about
    // verifying is not a verification command. Exercise: extract the
    // "- Verification: <command>" line's value.
    question(
      "how-to-verify", "How do I verify it?",
      entryPath ? lineMentioning(entryPath, "verif") : null,
      entryName ? `${entryName} (Verification line)` : null,
    ),
    // Naive draft: the file's first line, which is its heading.
    // Exercise: extract the "- Next best step: <text>" line's value.
    question(
      "where-are-we", "Where are we now?",
      rawFirstLine(join(repo, "claude-progress.md")),
      "claude-progress.md (Next best step line)",
    ),
  ];

  const answered = questions.filter((q) => q.answered).length;
  const total = questions.length;
  return {
    questions,
    answered,
    total,
    visibility_gap: total > 0 ? (total - answered) / total : 0,
    ready: answered === total,
  };
}

function main(argv: readonly string[]): number {
  const repo = argv[2];
  if (!repo || argv.length !== 3) {
    console.error("usage: main.ts <repo-dir>");
    return 2;
  }
  if (!existsSync(repo) || !statSync(repo).isDirectory()) {
    console.error(`error: not a directory: ${repo}`);
    return 2;
  }
  const report = readRepo(repo);
  process.stdout.write(JSON.stringify(report, null, 2) + "\n");
  return report.ready ? 0 : 1;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  process.exit(main(process.argv));
}

File: lectures/lecture-03-why-the-repository-must-become-the-system-of-record/exercises/exercise-01-fresh-session-answers/starter/python/main.py
"""fresh-session-answers exercise, Python starter.

All five questions are attempted, but three extractors are naive first
drafts with a realistic mistake each (see SPEC.md "Starter state"):
how-organized answers from the instructions file instead of the
architecture doc, how-to-verify grabs the first line that MENTIONS
verification instead of the Verification line, and where-are-we returns
the progress file's heading instead of the Next best step line. Fix the
three per SPEC.md. Run ../../verify.sh --stack=python until it exits 0.
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path


def first_prose_line(path: Path) -> str | None:
    if not path.is_file():
        return None
    for line in path.read_text(encoding="utf-8").split("\n"):
        stripped = line.strip()
        if stripped and not stripped.startswith("#"):
            return stripped
    return None


def line_mentioning(path: Path, needle: str) -> str | None:
    """Naive helper: first line whose text contains the needle (any case)."""
    if not path.is_file():
        return None
    for line in path.read_text(encoding="utf-8").split("\n"):
        if needle in line.lower() and line.strip():
            return line.strip()
    return None


def raw_first_line(path: Path) -> str | None:
    """Naive helper: first non-empty line, headings included."""
    if not path.is_file():
        return None
    for line in path.read_text(encoding="utf-8").split("\n"):
        if line.strip():
            return line.strip()
    return None


def tagged_line(path: Path, tag: str) -> str | None:
    if not path.is_file():
        return None
    match = re.search(
        rf"^- {re.escape(tag)}: (.+)$", path.read_text(encoding="utf-8"), re.MULTILINE
    )
    return match.group(1).strip() if match else None


def instructions_file(repo: Path) -> Path | None:
    for name in ("AGENTS.md", "CLAUDE.md"):
        if (repo / name).is_file():
            return repo / name
    return None


def read_repo(repo: Path) -> dict:
    entry = instructions_file(repo)

    def question(qid: str, text: str, answer: str | None, source: str | None) -> dict:
        return {
            "id": qid,
            "question": text,
            "answered": answer is not None,
            "answer": answer,
            "source": source if answer is not None else None,
        }

    questions = [
        question(
            "what-is-this", "What is this system?",
            first_prose_line(entry) if entry else None,
            entry.name if entry else None,
        ),
        # Naive draft: answers "how is it organized" from the instructions
        # file's overview line. Exercise: extract the first prose line of
        # docs/ARCHITECTURE.md, source "docs/ARCHITECTURE.md".
        question(
            "how-organized", "How is it organized?",
            first_prose_line(entry) if entry else None,
            entry.name if entry else None,
        ),
        question(
            "how-to-run", "How do I run it?",
            tagged_line(entry, "Run") if entry else None,
            f"{entry.name} (Run line)" if entry else None,
        ),
        # Naive draft: any line that mentions verification. Prose about
        # verifying is not a verification command. Exercise: extract the
        # "- Verification: <command>" line's value.
        question(
            "how-to-verify", "How do I verify it?",
            line_mentioning(entry, "verif") if entry else None,
            f"{entry.name} (Verification line)" if entry else None,
        ),
        # Naive draft: the file's first line, which is its heading.
        # Exercise: extract the "- Next best step: <text>" line's value.
        question(
            "where-are-we", "Where are we now?",
            raw_first_line(repo / "claude-progress.md"),
            "claude-progress.md (Next best step line)",
        ),
    ]

    answered = sum(1 for q in questions if q["answered"])
    total = len(questions)
    return {
        "questions": questions,
        "answered": answered,
        "total": total,
        "visibility_gap": (total - answered) / total if total else 0.0,
        "ready": answered == total,
    }


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print("usage: main.py <repo-dir>", file=sys.stderr)
        return 2
    repo = Path(argv[1])
    if not repo.is_dir():
        print(f"error: not a directory: {repo}", file=sys.stderr)
        return 2
    report = read_repo(repo)
    print(json.dumps(report, indent=2))
    return 0 if report["ready"] else 1


if __name__ == "__main__":
    sys.exit(main(sys.argv))
